ACCUMULO-2232 Added options to Combiner for handling deletes

keith-turner opened the pull request.

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java:
```diff
@@ -149,6 +159,37 @@ public void next() throws IOException {
 
   private Key workKey = new Key();
 
+  private static Cache<String,Long> loggedMsgCache = CacheBuilder.newBuilder().expireAfterWrite(1, TimeUnit.HOURS).build();
```
+final

joshelser replied:
Would be safer to include a high-watermark on the size of the cache. Maybe 500 elements just to prevent catastrophic failure if we somehow create a bug that causes things to be added to this cache incorrectly.

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java:
```diff
@@ -227,10 +271,25 @@ public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> op
+
+    isPartialCompaction = ((env.getIteratorScope() == IteratorScope.majc) && !env.isFullMajorCompaction());
+    if (options.containsKey(DELETE_HANDLING_ACTION_OPTION)) {
+      deleteHandlingAction = DeleteHandlingAction.valueOf(options.get(DELETE_HANDLING_ACTION_OPTION));
```
A warning message with the actual property being checked (DELETE_HANDLING_ACTION_OPTION) and the value that we have no enum-value for would be useful.

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java:
```diff
@@ -240,15 +299,18 @@ public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> op
     IteratorOptions io = new IteratorOptions("comb", "Combiners apply reduce functions to multiple versions of values with otherwise equal keys", null, null);
-    io.addNamedOption(ALL_OPTION, "set to true to apply Combiner to every column, otherwise leave blank. if true, " + COLUMNS_OPTION
-        + " option will be ignored.");
+    io.addNamedOption(ALL_OPTION,
```
Nit: unrelated formatting.

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java:
```diff
@@ -240,15 +299,18 @@ public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> op
+    io.addNamedOption(ALL_OPTION,
+        "set to true to apply Combiner to every column, otherwise leave blank. if true, " + COLUMNS_OPTION + " option will be ignored.");
     io.addNamedOption(COLUMNS_OPTION, "<col fam>[:<col qual>]{,<col fam>[:<col qual>]} escape non-alphanum chars using %<hex>.");
+    // TODO
```
Did you forget todo something?

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java:
```diff
@@ -275,6 +337,10 @@ public boolean validateOptions(Map<String,String> options) {
     }
 
+    if (options.containsKey(DELETE_HANDLING_ACTION_OPTION)) {
+      DeleteHandlingAction.valueOf(options.get(DELETE_HANDLING_ACTION_OPTION));
```
Same suggestion about a meaningful error message here.

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java:
```diff
@@ -313,4 +378,46 @@ public static void setColumns(IteratorSetting is, List<IteratorSetting.Column> c
+  }
+
+  /**
+   * Combiners may not work correctly with deletes. Sometimes when Accumulo compacts the files in a tablet, it only compacts a subset of the files. If a delete
```
Great writeup.

joshelser commented on core/src/test/java/org/apache/accumulo/core/iterators/user/CombinerTest.java:
```diff
@@ -786,4 +816,107 @@ public void testAdds() {
+      assertEquals(expected, readAll(ai));
+
+    } finally {
+      logger.removeAppender(appender);
```
Glad to see the try/finally logger reset.

joshelser commented on core/src/test/java/org/apache/accumulo/core/iterators/user/CombinerTest.java:
```diff
@@ -786,4 +816,107 @@ public void testAdds() {
+      runDeleteHandlingTest(input, expected, DeleteHandlingAction.THROW_EXCEPTION, paritalMajcIe);
+      Assert.fail();
+    } catch (IllegalStateException ise) {
+      Assert.assertTrue(ise.getMessage().contains("Saw a delete during a partial compaction"));
```
Try to include a meaningful error message if the assertion fails. We should be able to explain the exact reason the test failed just by looking at surefire output.

joshelser commented on core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java:
```diff
@@ -149,6 +159,37 @@ public void next() throws IOException {
+              public Long call() throws Exception {
+                log.error("Combiner of type " + this.getClass().getSimpleName()
+                    + " saw a delete during a partial compaction.  This could cause undesired results.  See ACCUMULO-2232.  Will not log subsequent occurences for at least 1 hour.");
+                return System.currentTimeMillis();
```
Do we really need the current time? Isn't `0` sufficient since we don't inspect the value of the cache entries? A comment that the value returned by the Callable is meaningless would also be nice.

joshelser commented:
As commented on JIRA before, approach looks good. Pointed out what I see as room for some improvements.

keith-turner commented:
@joshelser thanks for the feedback, it was very useful.   I think I fixed all the issues you found.

joshelser commented:
Great. I'll try to take another peek today, but if you think it's ready, don't feel obligated to wait on me.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java:
```diff
@@ -149,6 +160,38 @@ public void next() throws IOException {
+      switch (deleteHandlingAction) {
+        case LOG_ERROR:
+          try {
+            loggedMsgCache.get(this.getClass().getName(), new Callable<Long>() {
```
At first, I thought "Callable<Void>" might be better, but apparently CacheBuilder doesn't like that, so might be better to use "Callable<Boolean>", because it's more readable to see "Boolean.TRUE" inserted into the cache to denote that a log message has recently occurred. (Unless you were going to use the integer to denote how many times it has been logged before suppressing repeats after some threshold, like 5.)

keith-turner commented:
In 31fcf00 I changed the behaviour slightly.

I dropped the option to throw an exception.  I was thinking this option could prevent problems, but it can not.  It can still only detect problems.  For detecting problems I think logging an error is sufficient.

I also changed logging to log on partial and full major compactions.   If a full major compaction sees a delete, then its evidence that a problem could have occurred in the past with a partial compaction.

joshelser commented:
:+1:

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java:
```diff
@@ -313,4 +392,48 @@ public static void setColumns(IteratorSetting is, List<IteratorSetting.Column> c
   }
+
+  /**
+   * @since 1.6.4 1.7.1 1.8.0
```
Instead of offering the options `IGNORE` and `LOG_ERROR`, we should probably just always log a warning (if `REDUCE_ON_FULL_COMPACTION_ONLY` is not set) and let users disable it in the log4j config, rather than in code.

The `REDUCE_ON_FULL_COMPACTION_ONLY` should be treated as a distinct boolean value which can be set to `true` or `false` as a normal iterator option (similar to `Filter`'s `negate` option), with a default value of `false` to preserve existing behavior.

That would also minimize API changes, especially for 1.6.x and 1.7.x. While this isn't technically public API, it'd still be good to minimize API impact on users using those versions.

keith-turner replied:
I like this approach because it simplified things.  In bbea55d I got rid of the enum and swithced to a boolean.  There is still an addition of a public method to Combiner.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/iterators/Combiner.java:
```diff
@@ -313,4 +392,48 @@ public static void setColumns(IteratorSetting is, List<IteratorSetting.Column> c
+   * <p>
+   * For correctness deletes should not be used with columns that are combined OR the {@link DeleteHandlingAction#REDUCE_ON_FULL_COMPACTION_ONLY} option should
+   * be used. Only reducing on full major compactions may have negative performance implications.
+   *
```
Another option to give users the ability to "delete" without using problematic deletes is to give them the ability to return a `null` value, with the semantics of `null` being "drop them all". This, combined with the ability to detect which iterator scope and and major compaction mode is being run, could be used to drop data in a sensible way, suitable to the user's schema, without inserting problematic delete markers.

If we did that, it'd be good to add that here, with the recommendation to avoid using deletes entirely.

keith-turner replied:
I don't think this should be coupled with this fix.  Also not sure about doing this in 1.6.  I opened a new issue [ACCUMULO-4011](https://issues.apache.org/jira/browse/ACCUMULO-4011)
